src/cmd/services/m3coordinator/ingest/write.go

```diff
@@ -108,107 +115,148 @@ func (d *downsamplerAndWriter) Write(
-			// we are going to set a special tag that means the downsampler
-			// will write a graphite ID. This should really be plumbed
-			// through the downsampler in general, but right now the aggregator
-			// does not allow context to be attached to a metric so when it calls
-			// back the context is lost currently.
-			appender.AddTag(downsample.MetricsOptionIDSchemeTagName,
-				downsample.GraphiteIDSchemeTagValue)
-		}
+	defer appender.Finalize()
 
-		var appenderOpts downsample.SampleAppenderOptions
-		if downsampleOverride {
-			appenderOpts = downsample.SampleAppenderOptions{
-				Override: true,
-				OverrideRules: downsample.SamplesAppenderOverrideRules{
-					MappingRules: overrides.DownsampleMappingRules,
-				},
-			}
+	for _, tag := range tags.Tags {
+		appender.AddTag(tag.Name, tag.Value)
+	}
+
+	if tags.Opts.IDSchemeType() == models.TypeGraphite {
+		// NB(r): This is gross, but if this is a graphite metric then
+		// we are going to set a special tag that means the downsampler
+		// will write a graphite ID. This should really be plumbed
+		// through the downsampler in general, but right now the aggregator
+		// does not allow context to be attached to a metric so when it calls
+		// back the context is lost currently.
+		appender.AddTag(downsample.MetricsOptionIDSchemeTagName,
+			downsample.GraphiteIDSchemeTagValue)
+	}
+
+	var appenderOpts downsample.SampleAppenderOptions
+	if downsampleMappingRuleOverrides, ok := d.downsampleOverrideRules(overrides); ok {
+		appenderOpts = downsample.SampleAppenderOptions{
+			Override: true,
+			OverrideRules: downsample.SamplesAppenderOverrideRules{
+				MappingRules: downsampleMappingRuleOverrides,
+			},
 		}
+	}
 
-		samplesAppender, err := appender.SamplesAppender(appenderOpts)
+	samplesAppender, err := appender.SamplesAppender(appenderOpts)
+	if err != nil {
+		return err
+	}
+
+	for _, dp := range datapoints {
+		err := samplesAppender.AppendGaugeTimedSample(dp.Timestamp, dp.Value)
 		if err != nil {
 			return err
 		}
-
-		for _, dp := range datapoints {
-			err := samplesAppender.AppendGaugeTimedSample(dp.Timestamp, dp.Value)
-			if err != nil {
-				return err
-			}
-		}
 	}
 
 	return nil
 }
 
-func (d *downsamplerAndWriter) maybeWriteStorage(
+func (d *downsamplerAndWriter) writeToStorage(
 	ctx context.Context,
 	tags models.Tags,
 	datapoints ts.Datapoints,
 	unit xtime.Unit,
 	overrides WriteOptions,
 ) error {
-	var (
-		storageExists             = d.store != nil
-		useDefaultStoragePolicies = !overrides.WriteOverride
-	)
-
-	if !storageExists {
-		return nil
-	}
-
-	if storageExists && useDefaultStoragePolicies {
+	storagePolicies, ok := d.writeOverrideStoragePolicies(overrides)
+	if !ok {
 		return d.store.Write(ctx, &storage.WriteQuery{
 			Tags:       tags,
 			Datapoints: datapoints,
 			Unit:       unit,
-			Attributes: storage.Attributes{
-				MetricsType: storage.UnaggregatedMetricsType,
-			},
+			Attributes: storageAttriutesFromPolicy(unaggregatedStoragePolicy),
```

Comment:
nit: spelling `storageAttriutesFromPolicy` > `storageAttributesFromPolicy`

Reply:
And in a few other places.

Reply:
Good catch, TY.